pkg/apis/core/types_shootsnapshot.go

```diff
@@ -0,0 +1,60 @@
+package core
+
+import (
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// +genclient
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// ShootSnapshot contains a snapshot of the Shoot's state required to migrate the Shoot's control plane to a new Seed.
+type ShootSnapshot struct {
+	metav1.TypeMeta
+	// Standard object metadata.
+	metav1.ObjectMeta
+	// Specification of the ShootSnapshot.
+	Spec ShootSnapshotSpec
+}
```

Comment:
Why is it called `Snapshot`. For me snapshot would imply, that it is possible to take snapshots at any point in time (especially multiple ones). An later it is possible to revert to a dedicated snapshot. But this is definitely not meant here. Basically this resource just holds the actual shoot state.

Reply:
There is one small difference - the state is copied to the ShootSnapshot, it's not necessarily a direct result of a controller operating on a resource (this is why we kept the `status.state` in the extension resources and didn't change it to `status.snapshot` there as well).

Reply:
cc @swapnilgm wrt naming of the resource

Reply:
> For me snapshot would imply, that it is possible to take snapshots at any point in time (especially multiple ones). An later it is possible to revert to a dedicated snapshot. But this is definitely not meant here.

Agreed. Reverting to dedicated state is one of the meaning of snapshot. This might be the impression given by ShootSnapshot at first place. But look at this way, when it comes to migration, in some cases, we actually bring the live system down and restore it on new seed by reverting it from the  ShootSnapshot. We read this ShootSnapshot spec and try to reverts in this case. 

On other hand, `state` field in k8s is more like to filled field and we tries to match it for spec. And as per control loop philosophy, we interpret spec field and put the state there. But we never check the current state, load it and try to match to the spec/desired state.

There is very thin difference between interpreting last observed state as an "only snapshot" or as a "state". That's why agreed to keep the state field in extension specific resources as `status.state` and `status.snapshot`.

Reply:
@gardener/gardener-maintainers @swapnilgm @mandelsoft Let's finalise the naming convention. I'm slightly leaning towards `ShootSnapshot`

Pros and cons for both:
```
ShootSnapshot
Pros:
 - The state kept in the ShootSnapshot resource is copied from the extension resource. It’s not the same as a resource’s status field which is the direct result of a controller operating on that resource and trying reconcile a specific state
 - The state in ShootSnapshot is then used to restore a system.
Cons:
 - Implies that there are more than one ShootSnapshots and either one can be used to restore the state
```
```
ShootState
Pros:
 - There is only one state resource from which restoration can be done
 - Implies that it has the most recently copied state by the ShootState Sync Controller
Cons:
 - Implies that it is only used for monitoring the current state and not to restore it on a different seed
 - Slightly implies that it is the current state - due to network issues between seed and garden cluster this could not be the case
```

Reply:
I tried to put my interpretation of naming above. Of course such naming is perspective based. So, whatever is majority feels right go with it. My priority is to `ShootSnapshot`.  If majority think `ShootState` is correct then I’m fine with it. I don't have that strong opinion against `ShootState`.

Reply:
A remark to the cons list above: For monitoring I would expect ShootStatus, not state.

Reply:
I don't care at all. If you want to have a vote from me, then lets go with `ShootState`, backed by @mandelsoft's argument that any state in Kubernetes is just the observed at a given point in time and might already be outdated compared to the real world's state.

>  - Implies that it is only used for monitoring the current state and not to restore it on a different seed

I don't understand this. To me, this is not implied.

Reply:
> A remark to the cons list above: For monitoring I would expect ShootStatus, not state.

This conflicts somehow with the `.status` field of the `Shoot` resource. Let's go with `ShootState` and focus on the real value-adding implementations.